### PR TITLE
Internal: Support prop types in atomic widgets [EDS-386]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: core/base/module.php [hunk 1]
```diff
@@ -76,7 +76,7 @@ abstract public function get_name();
 	 * @access public
 	 * @static
 	 *
-	 * @return Module An instance of the class.
+	 * @return $this An instance of the class.
```
Comment (StyleShit) [inline core/base/module.php:79]: it's just for better type inference in the IDE

File: modules/atomic-widgets/prop-types/prop-type.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	 * @return array<string>
+	 */
+	public function get_dynamic_categories(): array {
+		return [];
+	}
+
+	public function jsonSerialize(): array {
+		return [
+			'type' => $this->get_type(),
+			'dynamicCategories' => $this->get_dynamic_categories(),
```
Comment (StyleShit) [inline modules/atomic-widgets/prop-types/prop-type.php:25]: should be snake case?

File: modules/atomic-widgets/module.php [hunk 3]
```diff
@@ -60,6 +76,28 @@ private function register_widgets( Widgets_Manager $widgets_manager ) {
 		$widgets_manager->register( new Atomic_Image() );
 	}
 
+	private function register_prop_types() {
+		// Primitive types.
+		$this->prop_types->register( new String_Type() );
```
Comment (blackcolt) [inline modules/atomic-widgets/module.php:81]: maybe register an array?

Comment (blackcolt) [inline modules/atomic-widgets/module.php:81]: maybe array? for multi register

**Review by Omerisra6 — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/prop-types/prop-type.php (hunk 2 above)
Comment (StyleShit) [reply modules/atomic-widgets/prop-types/prop-type.php:25]: Done

File: modules/atomic-widgets/module.php (hunk 3 above)
Comment (StyleShit) [reply modules/atomic-widgets/module.php:81]: no

File: modules/atomic-widgets/module.php [hunk 4]
```diff
@@ -23,20 +29,30 @@ class Module extends BaseModule {
...
 		$this->register_experiment();
 
 		if ( Plugin::$instance->experiments->is_feature_active( self::EXPERIMENT_NAME ) ) {
 			add_filter( 'elementor/editor/v2/packages', fn( $packages ) => $this->add_packages( $packages ) );
 			add_filter( 'elementor/widgets/register', fn( Widgets_Manager $widgets_manager ) => $this->register_widgets( $widgets_manager ) );
 
 			add_action( 'elementor/editor/after_enqueue_scripts', fn() => $this->enqueue_scripts() );
+
+			$this->register_prop_types();
+			add_filter( 'elementor/editor/localize_settings', fn( array $settings ) => $this->add_prop_types_config( $settings ) );
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply modules/atomic-widgets/module.php:55]: DOne

File: modules/atomic-widgets/module.php [hunk 5]
```diff
@@ -23,20 +29,30 @@ class Module extends BaseModule {
 		'editor-style',
 	];
 
+	/**
+	 * @var Prop_Types_Registry
+	 */
+	public $prop_types;
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply modules/atomic-widgets/module.php:35]: no
Will be hard to test

File: modules/atomic-widgets/prop-types/prop-types-registry.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		}
+
+		$this->types[ $type ] = $prop_type;
+	}
+
+	public function get( string $type ): ?Prop_Type {
+		return $this->types[ $type ] ?? null;
+	}
+
+	public function jsonSerialize(): array {
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply modules/atomic-widgets/prop-types/prop-types-registry.php:33]: Done

File: modules/atomic-widgets/base/atomic-widget-base.php [hunk 7]
```diff
@@ -149,15 +150,23 @@ private static function validate_schema( array $schema ) {
...
 
-		if ( ! $prop->validate( $prop->get_default() ) ) {
-			Utils::safe_throw( "Default value for `$key` prop is not of type `{$prop->get_type()}` in `{$widget_name}`." );
+		try {
+			$prop->validate( $prop->get_default() );
+		} catch ( \Exception $e ) {
+			Utils::safe_throw( "Default value for `$key` prop is invalid in `{$widget_name}` - {$e->getMessage()}" );
 		}
 
 		foreach ( $prop->get_constraints() as $constraint ) {
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply modules/atomic-widgets/base/atomic-widget-base.php:172]: Done

File: modules/atomic-widgets/compatibility.php [hunk 8]
```diff
@@ -14,7 +14,9 @@ public function register_hooks() {
 	}
 
 	private function add_atomic_dynamic_tags_settings( $settings ) {
-		$settings['atomicDynamicTags'] = $this->convert_dynamic_tags_to_atomic( $settings['dynamicTags']['tags'] );
+		if ( isset( $settings['dynamicTags']['tags'] ) ) {
```
Comment (Omerisra6) [inline modules/atomic-widgets/compatibility.php:17]: Move to another PR?

(I ❤️ linear b)

**Review by Omerisra6 — APPROVED** [review]

File: modules/atomic-widgets/compatibility.php (hunk 8 above)
Comment (StyleShit) [reply modules/atomic-widgets/compatibility.php:17]: https://github.com/elementor/elementor/pull/28501
